### PR TITLE
[lua] Loot table helper

### PR DESCRIPTION
**Pull request by MowFord**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Consolidates "loot group" table handling into a single place. There have been multiple bugs/edge cases caught and updated in `battlefield.lua`. This extends that treatment to anything else that uses the same type of loot selection with weights.
- assaults define an unappraised item for the box opener, as well as loot to the treasure pool
- garrison gives a random item from the level cap pool per player participating
  - Confirmed [here](https://youtu.be/mJcrQD9K6Qo?t=668) a party of 2 players get 2 different items, old lsb logic was giving the same item twice
- battlefields of course, though these are the only ones that can have GIL in the pool

The new util just takes a table of loot groups, which can have the optional `quantity` key, to define if the group should roll multiple times. It returns a table of selected loot, which then has to be distributed by whichever function called it. The complexity of handling all the different ways loot was distributed seemed better done in the various places (though it's usually just looping over the results and running `player:addTreasure(item)`)

Note: seagull grounded was reporting the wrong position on completion, so i updated to report F-11 instead of I-8

For sanity's sake, note that I did a mass find replace in `scripts/battlefields`, so reviewing everything minus that last commit might be helpful: https://github.com/LandSandBoat/server/compare/b8e48fd20615c439a5d2e59ab7dc0d7a23b337ec..b9d193f6fafab8b0f98f514200255faaa7fef8d2

## Steps to test these changes

- assaults
  - kind of a pain because of the prereqs, so i'll outline those first
    - go to whitegate, `!gotoname Lageegee`
    - give yourself the prereq to talk to him: `!addkeyitem IMPERIAL_ARMY_ID_TAG`
    - choose the seagull mission
  - go to mire, then `!gotoname _273` and enter seagull grounded
    - if you fail the mission, to restart without going back to whitegate
      - `!setplayervar <name> assaultEntered 0`
    - `!addkeyitem assault_armband`
    - `!exec player:getInstance():complete()`
    - `!gotoname Ancient_Lockbox`
    - see you get an unappraised item as well as the pool gets some treasure
    - image
- [Garrison](https://www.bg-wiki.com/ffxi/Category:Garrison)
  - `!additem bunny_fang_sack`
  - go to cape Terrigan
  - magic the zone into your nation's control?
    - I had to rip out the `xi.garrison.getAllyInfo` check in `xi.garrison.onTrade` and set nationID to zero explicitly in `xi.garrison.spawnNPCs`
      - recommend setting `enableDebugPrints = true` at the top to know how to get things moving
    - then show the npc: `!exec GetNPCByID(17240473):setStatus(0)`
  - and trade the `!additem bunny_fang_sack`
  - murder the waves and `!immortal` at least one npc helper so you don't fail
  - see lewtz
  - image
- battlefields, since the function was basically copy/pasted from this one, shouldn't be worrisome, but needs testing nonetheless
  - go to throne room zone
  - `!additem moon_orb`
  - enter kindred spirits
  - kill the mobs
  - see lewtz
  - image

### CODE REVIEW

File: scripts/globals/garrison_data.lua [hunk 1]
```diff
@@ -163,65 +163,65 @@ xi.garrison.loot =
...
+        { item = xi.item.MIGHTY_TALWAR,     weight =  350 },
+        { item = xi.item.RAI_KUNIMITSU,     weight =  350 },
+        { item = xi.item.NUKEMARU,          weight =  350 },
+        { item = xi.item.MIGHTY_PICK,       weight =  350 },
+        { item = xi.item.MIGHTY_KNIFE,      weight =  350 },
+        { item = xi.item.MIGHTY_ZAGHNAL,    weight =  350 },
+        { item = xi.item.MIGHTY_LANCE,      weight =  350 },
+        { item = xi.item.MIGHTY_AXE,        weight =  350 },
+        { item = xi.item.MIGHTY_PATAS,      weight =  350 },
+        { item = xi.item.MIGHTY_SWORD,      weight =  350 },
```
Comment (Xaver-DaRed) [inline scripts/globals/garrison_data.lua]: Please rename `item` to `itemId`
The original name wasnt proper either.

File: scripts/globals/utils.lua [hunk 2]
```diff
@@ -1329,3 +1329,72 @@ function utils.timeIsBefore(current, target)
...
+    for _, lootGroup in ipairs(lootTable) do
+        local max = 0
+
+        -- not ipairs because we might have string keys alongside the individual items in each lootGroup
+        for j, entry in pairs(lootGroup) do
+            if type(entry) == 'table' then
+                max = max + entry.weight
+
+                if entry.item == nil then
+                    print(fmt('[ERROR] Player ({}) has encountered nil item at index {} of lootGroup with index {}', actor:getName(), j, i))
```
Comment (Xaver-DaRed) [inline scripts/globals/utils.lua:1373]: `i`?

File: scripts/globals/garrison_data.lua (hunk 1 above)
Comment (MowFord) [reply scripts/globals/garrison_data.lua]: I took the path of least resistance, that would require updating every battlefield. I'll do it if you still choose that route just making sure

File: scripts/globals/utils.lua [hunk 3]
```diff
@@ -1329,3 +1329,72 @@ function utils.timeIsBefore(current, target)
...
+    if not actor or not actor:getName() then
+        return selectedLoot
+    end
+
+    for _, lootGroup in ipairs(lootTable) do
+        local max = 0
+
+        -- not ipairs because we might have string keys alongside the individual items in each lootGroup
+        for j, entry in pairs(lootGroup) do
+            if type(entry) == 'table' then
```
Comment (Xaver-DaRed) [inline scripts/globals/utils.lua:1369]: It should always be a table, so remove this check and let it break if it isnt

File: scripts/globals/utils.lua (hunk 2 above)
Comment (MowFord) [reply scripts/globals/utils.lua:1373]: fixed, i cleaned up the loop from `for i = 1, #lootGroups` to using ipairs, forgot about the print using index

File: scripts/globals/utils.lua [hunk 4]
```diff
@@ -1329,3 +1329,72 @@ function utils.timeIsBefore(current, target)
...
+        end
+
+        local quantity = lootGroup.quantity or 1
+
+        for j = 1, quantity do
+            local roll    = math.random(max)
+            local current = 0
+
+            for _, entry in pairs(lootGroup) do
+                if type(entry) == 'table' then
```
Comment (Xaver-DaRed) [inline scripts/globals/utils.lua:1385]: Same here

File: scripts/globals/utils.lua (hunk 3 above)
Comment (MowFord) [reply scripts/globals/utils.lua:1369]: lootGroups are allowed to define `quantity` which determines how many times to loop the group, so the entries are either a table with no key (auto indexed) or a string key `lootGroup.quantity`

File: scripts/zones/Periqia/npcs/Ancient_Lockbox.lua [hunk 5]
```diff
@@ -42,57 +42,57 @@ entity.onTrigger = function(player, npc)
...
-                { itemid = xi.item.HI_POTION_TANK,    droprate = 400 },
-                { itemid = 0,                         droprate = 600 },
+                { item = xi.item.HI_POTION_TANK,    weight = 400 },
+                { item = 0,                         weight = 600 },
             },
             {
-                { itemid = xi.item.HI_RERAISER,       droprate = 200 },
-                { itemid = 0,                         droprate = 800 },
+                { item = xi.item.HI_RERAISER,       weight = 200 },
+                { item = 0,                         weight = 800 },
```
Comment (Xaver-DaRed) [inline scripts/zones/Periqia/npcs/Ancient_Lockbox.lua]: Same as my original `itemId` comment, but this time applied to all of them

File: scripts/globals/garrison_data.lua (hunk 1 above)
Comment (Xaver-DaRed) [reply scripts/globals/garrison_data.lua]: an item can be a global object, just like skill. itemId, just like skillId is exactly that: A pointer/enum.

We should try to stick to naming things what they actually are, even if historically everyone has named things however they've wanted.

Comment (MowFord) [reply scripts/globals/garrison_data.lua]: sounds good, will take me a bit but i'll get it done

Comment (MowFord) [reply scripts/globals/garrison_data.lua]: Ok done and re-tested. Found a stray comment on the util that was no longer true, and updated the example lootTable to include a gil row

Comment (MowFord) [reply scripts/globals/garrison_data.lua]: done round 2 with proper `itemId`

**Review by Xaver-DaRed — APPROVED** [review]
